File: JavaSlideshow/app/src/main/java/com/mobappdev/javaslideshow/service/Slideshow.java
package com.mobappdev.javaslideshow.service;

import android.util.Log;

import com.mobappdev.javaslideshow.model.Slide;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class Slideshow {

    private List<Slide> slides = new ArrayList<>();
    public String sortDirection = "asc";
    private static Slideshow slideshow;

    public static Slideshow getInstance() {
        if (Slideshow.slideshow == null) {
            Slideshow.slideshow = new Slideshow();
        }
        return Slideshow.slideshow;
    }

    public void addSlide(Slide newSlide) {
        slides.add(newSlide);
    }

    public void shuffle() {
        Collections.shuffle(slides);
    }

    public Slide next(int counter){
        if(counter < slides.size()){
            return slides.get(counter);
        }else{
            return slides.get(1);
        }
    }

    public int getTotalSlides() {
        return slides.size();
    }

    public void sort() {
        if (sortDirection.equalsIgnoreCase("desc")) {
            Collections.reverse(slides);
            sortDirection = "asc";
        } else if (sortDirection.equalsIgnoreCase("asc")) {
            Collections.sort(slides, new SlideComparator());
            sortDirection = "desc";
        }
    }

    public class SlideComparator implements Comparator<Slide> {
        @Override
        public int compare(Slide o1, Slide o2) {
            int value1 = o1.getTimestamp().compareTo(o2.getTimestamp());
            if (value1 == 0) {
                int value2 = o1.getTitle().compareTo(o2.getTitle());
                if (value2 == 0) {
                    return o1.getId() - o2.getId();
                } else {
                    return value2;
                }
            }
            return value1;
        }
    }

}

File: KotlinSlideshow/app/src/main/java/at/fhj/iit/slideshowgroupa/service/Slideshow.kt
package at.fhj.iit.slideshowgroupa.service

import at.fhj.iit.slideshowgroupa.model.Slide

// object = singleton class
object Slideshow {
    private var slides = mutableListOf<Slide>()
    var sortDirection = "asc"

    fun addSlide(aNewSlide:Slide){
        slides.add(aNewSlide)
    }

    override fun toString():String{
        val sortedSlides = slides.sortedBy { it.id }
        return sortedSlides.joinToString(" and ", "Slides: ")
    }

   fun shuffle(){
        slides.shuffle()
    }

    fun next(counter:Int):Slide{
        return slides[counter]
    }

    fun getTotalSlides():Int{
        return slides.size
    }


    fun sort(){
        if(sortDirection == "desc"){
            slides.reverse()
            sortDirection = "asc"
        } else if (sortDirection === "asc") {
            slides.sortWith(compareBy({it.timestamp}, {it.title}, {it.id}))
            sortDirection="desc"
        }
    }


}